File: 0355-design-twitter/0355-design-twitter.cpp
class Twitter {
public:
    unordered_map<int,set<int>>followers;
    unordered_map<int,vector<pair<int,int>>>tweets;
    long long time;
    Twitter() {
        time=0;
    }
    
    void postTweet(int userId, int tweetId) {
        tweets[userId].push_back({time++,tweetId});
    }
    
    vector<int> getNewsFeed(int userId) {
        priority_queue<pair<int,int>>maxHeap;
        for(auto tweet:tweets[userId]){
            maxHeap.push(tweet); //tweets and timestamps
        }
        for(auto f:followers[userId]){
            for(auto p2:tweets[f])
              maxHeap.push(p2);
        }
        vector<int>res;
        while(res.size() < 10 && maxHeap.size() >0){
            res.push_back(maxHeap.top().second);
            maxHeap.pop();
        }
        return res;

    }
    
    void follow(int followerId, int followeeId) {
        followers[followerId].insert(followeeId);
    }
    
    void unfollow(int followerId, int followeeId) {
        followers[followerId].erase(followeeId);
    }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */

File: 0355-design-twitter/0355-design-twitter.java
class Twitter {

    private static int timestamp = 0;

    private static class Tweet {
        int tweetId;
        int time;

        Tweet(int tweetId, int time) {
            this.tweetId = tweetId;
            this.time = time;
        }
    }

    private Map<Integer, Set<Integer>> follows;
    private Map<Integer, List<Tweet>> tweets;

    public Twitter() {
        follows = new HashMap<>();
        tweets = new HashMap<>();
    }

    public void postTweet(int userId, int tweetId) {
        tweets.putIfAbsent(userId, new ArrayList<>());
        tweets.get(userId).add(new Tweet(tweetId, timestamp++));
    }

    public List<Integer> getNewsFeed(int userId) {
        PriorityQueue<Tweet> pq = new PriorityQueue<>((a, b) -> b.time - a.time);

        // Ensure the user follows themselves
        follows.putIfAbsent(userId, new HashSet<>());
        follows.get(userId).add(userId);

        for (int followeeId : follows.get(userId)) {
            List<Tweet> tweetList = tweets.getOrDefault(followeeId, new ArrayList<>());
            for (int i = tweetList.size() - 1; i >= 0 ; i--) {
                pq.offer(tweetList.get(i));
            }
        }

        List<Integer> feed = new ArrayList<>();
        int count = 0;
        while (!pq.isEmpty() && count < 10) {
            feed.add(pq.poll().tweetId);
            count++;
        }
        return feed;
    }

    public void follow(int followerId, int followeeId) {
        follows.putIfAbsent(followerId, new HashSet<>());
        follows.get(followerId).add(followeeId);
    }

    public void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId) return; // Can't unfollow yourself
        if (follows.containsKey(followerId)) {
            follows.get(followerId).remove(followeeId);
        }
    }
}


/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */